docs/rfc/block-registration.md

```diff
@@ -0,0 +1,369 @@
+This RFC is intended to serve both as a specification and as documentation for the implementation of runtime-agnostic block type registration.
+
+## Requirements
+
+Behind any block type registration is some abstract concept of a unit of content. This content type can be described without consideration of any particular technology. In much the same way, we should be able to describe the core constructs of a block type in a way which can be interpreted in any runtime.
+
+In more practical terms, an implementation should fulfill requirements that...
+
+* A block type registration should be declarative and context-agnostic. Any runtime (PHP, JS, or other) should be able to interpret the basics of a block type (see "Block API" in the sections below) and should be able to fetch or retrieve the definitions of the context-specific implementation details. The following things should be made possible:
+  * Building REST APIs to fetch the available block types
+  * Building REST APIs to fetch block objects from posts
+* This API should be backward compatible with what we have at the moment.
+* It should be possible to statically analyze a block type in order to support advanced use-cases required by one of the [9 projects](https://make.wordpress.org/core/2018/12/08/9-priorities-for-2019/) for 2019 in Wordpress: "Building a WordPress.org directory for discovering blocks, and a way to seamlessly install them."
```

Comment:
> This could use a great deal more clarification

The block repository/directory should not have to parse JavaScript files or php files to understand the block types, their names, icons, descriptions... in order to show a proper directory.

It can statically analyze the files of any plugin to retrieve blocks and their properties.

Reply:
Fixed `p` in 261754e and clarified directory thing in b8d0e1b.